cpp/src/TrinoExchangeSource.cpp

```diff
@@ -148,74 +178,86 @@ void TrinoExchangeSource::processDataResponse(
                                  .getSingleOrEmpty(protocol::TRINO_PAGE_NEXT_TOKEN_HEADER)
                                  .c_str());
 
-  std::unique_ptr<exec::SerializedPage> page;
-  std::unique_ptr<folly::IOBuf> singleChain;
+  std::vector<std::unique_ptr<exec::SerializedPage>> pages;
+  pages.reserve(ackSequence - sequence_ + (complete ? 1 : 0));
   const bool empty = response->empty();
   int64_t totalBytes{0};
   if (!empty) {
     auto iobufs = response->consumeBody();
-    // trino have a 16 bytes header, including 4 bytes magic, 8 bytes checksum, 4 bytes
-    // page size
-    // Refer to io.trino.operator.HttpPageBufferClient.PageResponseHandler#handle
-    // We need skip only once.
-    for (auto& buf : iobufs) {
-      totalBytes += buf->capacity();
-      if (!singleChain) {
-        buf->trimStart(16);
-        singleChain = std::move(buf);
-      } else {
-        singleChain->prev()->appendChain(std::move(buf));
-      }
+    // TODO: Update memory
+    // TrinoExchangeSource::updateMemoryUsage(totalBytes);
+    auto firstBuf = iobufs.front();
+    VELOX_CHECK_GE(firstBuf->length(), TRINO_RESULT_RESPONSE_HEADER_LEN);
+
+    const int32_t magic = *reinterpret_cast<const int32_t*>(firstBuf->data());
+    const int64_t checkSum = *reinterpret_cast<const int64_t*>(firstBuf->data() + 4);
+    const int32_t pageCount = *reinterpret_cast<const int32_t*>(firstBuf->data() + 12);
+
+    VELOX_CHECK_EQ(magic, TRINO_RESULT_RESPONSE_HEADER_MAGIC);
+    VELOX_CHECK_EQ(sequence_ + pageCount, ackSequence);
+    iobufs.trimStart(TRINO_RESULT_RESPONSE_HEADER_LEN);
+
+    int32_t remaingPage = pageCount;
+    int8_t pageHeader[TRINO_SERIALIZED_PAGE_HEADER_SIZE + 4];
+    while (remaingPage) {
+      readIOBuf(pageHeader, iobufs.front(), TRINO_SERIALIZED_PAGE_HEADER_SIZE + 4);
```

Comment:
Do we really need readIObuf here? I believe we just need to know the compressed size here.

Reply:
Because the bytes of compressed size may be placed on two IOBuf nodes, the memory is not successive.

Reply:
Yes. So could we return a compressed size from this function instead of doing `memcpy`?

Reply:
Or we could write a simple loop to move 17 bytes, to avoid the cost of ```memcpy```

Reply:
Besides, why is it 17 bytes? 4 for row numbers, 1 for codec, 4 for uncompressed size, 4 for compressed size, what's the extra 4 bytes?

Reply:
It's the column's number? I don't think we need those information here.

Reply:
Ok, remove them here and log them in the deserializer.